sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/ReadTask.java

```diff
@@ -37,13 +37,19 @@
    * The preferred locations where this read task can run faster, but Spark does not guarantee that
    * this task will always run on these locations. The implementations should make sure that it can
    * be run on any location. The location is a string representing the host name of an executor.
+   *
+   * If an exception was thrown, the action would fail and we guarantee that no Spark job was
+   * submitted.
    */
   default String[] preferredLocations() {
     return new String[0];
```

Comment:
Somewhere in the spark code there's that translation of localhost -> anywhere, so that things like object stores who return "localhost" end up with independent-of-location placement. Is this required to have happened by the time `preferredLocations` is called, or can an impl return "localhost" and expect spark to deal with it. Purest would be to say "stores which use 'localhost' as a hint to mean 'unplaced' are required to have filtered it out by this point'

Reply:
for `independent-of-location placement`, users can just return `new String[0]`.

I don't know where Spark treats "localhost" as "anywhere", but I'd expect it to be an unrecognized host and then it's "anywhere".

Reply:
I'm sure there was some specific filter for it, though a quick grep only shows that happening in {{ReliableCheckpointRDD}}. The reason the filter is needed is that getHostByName(localhost) does return a host, but scheduling gets confused, without the filtering work can get get held back until the driver concludes that localhost isn't free & so assign it elsewhere in the cluster (Hive can do this unintentionally, which is why I did trace through spark's use of getBlockLocations once)